Replace platform support with plain application selection during submission

EnTeQuAk opened the pull request:
This pull request removes supported platform selection and replaces it
with selecting application explicitly. This way the developer can choose
between "Firefox" and "Firefox for Android" support and platforms are
always naively set to "PLATFORMS_ALL" since we want to get rid of
platforms eventually.

* This cleans up most of the platform selection code that resides in devhub, views and templates since it's essentially unused now
* This doesn't touch any of the platform handling in models since all of that has to continue working for legacy add-ons and WebExtensions that have different platforms selected
* We are defaulting to "Firefox" (Desktop) compatibility with Android compatibility being optional

Regarding static themes, they are only compatible with Firefox Desktop currently which works seemlessly since we are defaulting to Firefox Desktop compatibility by default anyway and you can't select explicit Android compatibility for static themes.

That's why I removed all that "platforms to exclude" code and didn't implement a replacement for compatibility apps, any objections to that? We can do that if we have to, e.g if we decide to default to Firefox and Android support for regular WebExtensions.

Fixes #8752
Fixes #6067

This allows selecting support for an application directly during submission

New add-on submission:
![screenshot_2018-09-07 upload add-on developer hub add-ons for firefox](https://user-images.githubusercontent.com/139033/45202820-7c8a6900-b27a-11e8-96b1-66bf334f57fd.png)

New version submission:
![screenshot_2018-09-07 upload add-on spird-lapwine-reingonbis add-ons for firefox](https://user-images.githubusercontent.com/139033/45202890-aba0da80-b27a-11e8-85ec-f04b8bdebe4a.png)


TODO:

* [x] Clean up and add more tests for the submission step, fix failing tests
* [x] More testing, more clicking through
* [x] Test submitting addon with all apps
* [x] Test submitting addon only with Firefox
* [x] Test submitting addon still respects strict min/max versions from manifest
* [x] Test static theme submission (since it's re-using the same form)

EnTeQuAk commented:
@jvillalobos looking at https://github.com/mozilla/addons-server/issues/6576#issuecomment-362808948, do you still want android to be switched off by default? This PR assumes that it's on by default but it's easy to change

diox commented on src/olympia/addons/models.py:
```diff
@@ -591,8 +591,9 @@ def initialize_addon_from_upload(cls, data, upload, channel, user):
+                    user=None):
         """
         Create an Addon instance, a Version and corresponding File(s) from a
         FileUpload, a list of platform ids, a channel id and the
```
heads up: The docstring will need to be updated

EnTeQuAk replied:
yeah, once I'm done with the first wave of changes I'll do a whole-code sweep for anything `platform` related I may have missed :)

Thanks!

jvillalobos commented:
Yes, let's default to Firefox support only.

EnTeQuAk commented:
This may still be quite naive since it's the first time I'm touching most of this submission related code, so r? @diox and @eviljeff because more input will certainly help to get this right.

EnTeQuAk commented on src/olympia/addons/tasks.py:
```diff
@@ -720,7 +722,7 @@ def migrate_legacy_dictionary_to_webextension(addon):
     # Create version.
     version = Version.from_upload(
-        upload, addon, platforms=[amo.PLATFORM_ALL.id],
+        upload, addon, selected_apps=[amo.FIREFOX.id],
```
@diox does that work? Do you know if WebExtension dictionaries are compatible with Android?

diox replied:
Yup

diox commented on src/olympia/devhub/views.py:
```diff
@@ -1335,26 +1320,25 @@ def _submit_upload(request, addon, channel, next_view, version=None,
+                upload=data['upload'],
+                version=version,
+                platform=amo.PLATFORM_ALL.id,
+                parsed_data=data['parsed_data'])
```
We should prevent file uploads (any developer affected by this restriction on their existing versions can just upload a new one) and preventing changing platform on existing files in this PR.

EnTeQuAk replied:
As discussed in IRC, agreed. I didn't think of separate file uploads, I'll disable all that.

EnTeQuAk commented on src/olympia/devhub/forms.py:
```diff
@@ -477,34 +492,30 @@ class NewUploadForm(forms.Form):
+        choices=amo.APPS_FIREFOXES_ONLY_CHOICES,
+        # Pre-select only Desktop Firefox, most of the times developers
+        # don't develop their WebExtensions for Android.
+        # See this GitHub comment: https://bit.ly/2QaMicU
```
too long, really doesn't fit into the line length :( I'll see what I can doo

diox replied:
"See comment by jvillabos in https://github.com/mozilla/addons-server/issues/6576" fits on 2 lines :)

EnTeQuAk commented on src/olympia/devhub/forms.py:
```diff
@@ -477,34 +492,30 @@ class NewUploadForm(forms.Form):
 
-        # If we have a version reset platform choices to just those compatible.
+        # If we have a version, pre-select the ones from the previous
+        # version to help the developer decide.
```
true, I guess I was thinking about "this is the new version, the other one is the previous" but given the actual implementation I'll change it to "current version"

EnTeQuAk commented:
Removed platform editing from files as discussed.

![screenshot from 2018-09-17 09-58-20](https://user-images.githubusercontent.com/139033/45612110-34312f00-ba62-11e8-8b98-7311b2afb79b.png)

The platforms are still visible, although read-only right now.

EnTeQuAk commented:
Now I also removed explicit file uploading as discussed in IRC and here since it's not shown for new uploads anyway and developers wanting to change files should just upload a new version.

eviljeff commented on src/olympia/addons/tasks.py:
```diff
@@ -721,6 +721,8 @@ def migrate_legacy_dictionary_to_webextension(addon):
 
     parsed_data = parse_addon(upload, user=user)
     # Create version.
+    # WebExtension dictionaries are only compatible with Firefox
```
nit: "Firefox" is ambiguous - "Firefox Desktop" is clearer

eviljeff commented on src/olympia/devhub/forms.py:
```diff
@@ -508,8 +508,8 @@ def __init__(self, *args, **kw):
-        # version to help the developer decide.
+        # If we have a version, pre-select the ones from the current version
+        # to help the developer decide.
         if self.version:
```
You can drop everything `self.version` now, right?  Because you can't have any new uploads for a version once it exists.

EnTeQuAk replied:
True, I think I was too cautious. I'll get rid of all that.

eviljeff commented on src/olympia/devhub/views.py:
```diff
@@ -1424,14 +1425,6 @@ def submit_version_theme_wizard(request, addon_id, addon, channel):
-    version = get_object_or_404(Version, id=version_id)
-    return _submit_upload(
-        request, addon, version.channel, 'devhub.submit.file.finish',
-        version=version)
```
Remove support for `version` parameter from `_submit_upload` too.

eviljeff commented on src/olympia/devhub/views.py:
```diff
@@ -1620,12 +1613,6 @@ def submit_version_finish(request, addon_id, addon, version_id):
-@dev_required
-def submit_file_finish(request, addon_id, addon, version_id):
-    version = get_object_or_404(Version, id=version_id)
-    return _submit_finish(request, addon, version, is_file=True)
```
and remove the `is_file` parameter from `_submit_finish`

EnTeQuAk commented:
> just a bit more red to go

I guess I was too cautious, I removed all the mentioned code as well and also removed `submit_page=file` support from our templates since that won't happen anymore either.

eviljeff commented on src/olympia/devhub/views.py:
```diff
@@ -1302,32 +1302,22 @@ def submit_version_distribution(request, addon_id, addon):
-    version will need both an addon and a version supplied.
+def _submit_upload(request, addon, channel, next_view, wizard=False):
+    """ If this is a new addon upload `addon` will be None;
+    if this is a new version upload `version` will be None.
```
nit: comment still mentions `version`

EnTeQuAk commented:
Rebased to master, if there are no objections I'm going to merge this into master once tests are green. This is easy enough to revert in case of second-thoughts since no database-data is being changed.

diox commented on src/olympia/devhub/forms.py:
```diff
@@ -477,35 +492,21 @@ class NewUploadForm(forms.Form):
-        error_messages={'required': 'Need at least one platform.'}
-    )
+        widget=CompatAppSelectWidget(),
+        error_messages={'required': _('Need to select at least one app.')})
```
"application" and not "app", I think.
